flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/operation/OperationManager.java

```diff
@@ -138,21 +153,188 @@ public ResultSet fetchResults(OperationHandle operationHandle, long token, int m
+        }
+
+        void runBefore() {
+            updateState(OperationStatus.RUNNING);
+        }
+
+        void runAfter() {
+            updateState(OperationStatus.FINISHED);
+        }
+
+        public void run() {
+            try {
+                operationLock.acquire();
+                LOG.debug(
+                        String.format(
+                                "Operation %s acquires the operation lock.", operationHandle));
+                updateState(OperationStatus.PENDING);
+                Runnable work =
+                        () -> {
+                            try {
+                                runBefore();
+                                resultFetcher = resultSupplier.call();
+                                runAfter();
+                            } catch (Throwable t) {
+                                processThrowable(t);
+                            }
+                        };
+                // Please be careful: the returned future by the ExecutorService will not wrap the
+                // done method.
+                FutureTask<Void> copiedTask =
+                        new FutureTask<Void>(work, null) {
+                            @Override
+                            protected void done() {
+                                LOG.debug(
+                                        String.format(
+                                                "Release the operation lock: %s when task completes.",
+                                                operationHandle));
+                                operationLock.release();
+                            }
+                        };
+                service.submit(copiedTask);
+                invocation = copiedTask;
+                // If it is canceled or closed, terminate the invocation.
+                OperationStatus current = status.get();
+                if (current == OperationStatus.CLOSED || current == OperationStatus.CANCELED) {
+                    LOG.debug(
+                            String.format(
+                                    "The current status is %s after updating the operation %s status to %s. Close the resources.",
+                                    current, operationHandle, OperationStatus.PENDING));
+                    closeResources();
+                }
+            } catch (Throwable t) {
+                processThrowable(t);
+                throw new SqlGatewayException(
+                        "Failed to submit the operation to the thread pool.", t);
+            } finally {
+                if (invocation == null) {
+                    // failed to submit to the thread pool and release the lock.
+                    LOG.debug(
+                            String.format(
+                                    "Operation %s releases the operation lock when failed to submit the operation to the pool.",
+                                    operationHandle));
+                    operationLock.release();
```

Comment:
The operationLock will also be released in copiedTask, which then cause release for twice.

Reply:
When enter to here, it means the Operation fails to submit itself to the thread pool and the copiedTask will not execute.